# Import libraries and dataset

In [3]:
import pandas as pd
import numpy as np 
import os 

path = r'C:\Users\ctede\OneDrive\Desktop\Instacart Basket Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_new_columns.pkl'))

In [6]:
#Create a subset df of the first one million entries
df = ords_prods_merge[:1000000]

In [8]:
df.shape

(1000000, 18)

In [9]:
df.head()

order_id  customer_id  order_number order_dow  order_hour_of_day  \
0  2539329            1             1         2                  8   
1  2398795            1             2         3                  7   
2   473747            1             3         3                 12   
3  2254736            1             4         4                  7   
4   431534            1             5         4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     0.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day new_busiest_days busiest_period_of_day  
0  Regularly busy   Regularly busy        Average orders  
1  Regularly busy     Slowest days        Average orders  
2  Regularly busy     Slowest days           Most orders  
3      Least busy     Slowest days        Average orders  
4      Least busy     Slowest days           Most orders

# Group by 'product_name' column

In [10]:
df.groupby('product_name')

In [11]:
#Creates a pandas object (indicated by the term "object"). The output of the groupby function is not visible. 
#You still need to aggregate the data or apply a function before you can see the results
#By itself, groupby() is worthless in terms of output 

# Using the agg() function 

In [12]:
#Produce single descriptive statistic for "order_number" and group by "department_id" 
df.groupby('department_id').agg({'order_number' : ['mean']})
#Groupby function is being assigned to the df dataframe and creates a pandas object for "department_id". The agg() function
#will return the mean of the given column "order_number"

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

### Table that lists all the possible "department_id" values with their corresponding "order_number" means/averages. = The average number of orders per department ID. i.e. The product (4) department has a mean of around 19 (18.8). The product department sells considerably more on average than household goods (17). 

In [13]:
#Alternative method to group the department id with the average/mean order numbers 
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [14]:
#agg() --> the columns you want to aggregeate go inside the parentheses of the agg() function
#mean() --> index the column with square brackets, then follow the function you want to use after the dot
#Square brackets are only used for indexing 

In [15]:
#Produce multiple statistics at the same time on "order_number" column 
df.groupby('department_id').agg({'order_number':['mean','min','max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

# Using the transform() function

In [17]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['customer_id'])['order_number'].transform(np.max)

In [18]:
#'max_order' new column is created and this is where the maximum order number for each user is stored. 
#the dataframe is grouped by the "customer_id" column
#the transform() function is applied on the "order_number" column with the np.max argument to find the max value
#(from the NumPy library)

In [19]:
ords_prods_merge.head(15)

order_id  customer_id  order_number order_dow  order_hour_of_day  \
0   2539329            1             1         2                  8   
1   2398795            1             2         3                  7   
2    473747            1             3         3                 12   
3   2254736            1             4         4                  7   
4    431534            1             5         4                 15   
5   3367565            1             6         2                  7   
6    550135            1             7         1                  9   
7   3108588            1             8         1                 14   
8   2295261            1             9         1                 16   
9   2550362            1            10         4                  8   
10  2968173           15            15         1                  9   
11  1870022           15            17         2                 16   
12  1911383           15            18         2                 11   
13  2715276           15            21         1                  9   
14   487368           15            22         1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                      0.0         196                  1          0   both   
1                     15.0         196                  1          1   both   
2                     21.0         196                  1          1   both   
3                     29.0         196                  1          1   both   
4                     28.0         196                  1          1   both   
5                     19.0         196                  1          1   both   
6                     20.0         196                  1          1   both   
7                     14.0         196                  2          1   both   
8                      0.0         196                  4          1   both   
9                     30.0         196                  1          1   both   
10                     7.0         196                  2          0   both   
11                     8.0         196                  6          1   both   
12                     7.0         196                  1          1   both   
13                     7.0         196                  1          1   both   
14                    14.0         196                  1          1   both   

   product_name  aisle_id  department_id  prices    price_range_loc  \
0          Soda        77              7     9.0  Mid-range product   
1          Soda        77              7     9.0  Mid-range product   
2          Soda        77              7     9.0  Mid-range product   
3          Soda        77              7     9.0  Mid-range product   
4          Soda        77              7     9.0  Mid-range product   
5          Soda        77              7     9.0  Mid-range product   
6          Soda        77              7     9.0  Mid-range product   
7          Soda        77              7     9.0  Mid-range product   
8          Soda        77              7     9.0  Mid-range product   
9          Soda        77              7     9.0  Mid-range product   
10         Soda        77              7     9.0  Mid-range product   
11         Soda        77              7     9.0  Mid-range product   
12         Soda        77              7     9.0  Mid-range product   
13         Soda        77              7     9.0  Mid-range product   
14         Soda        77              7     9.0  Mid-range product   

       busiest_day new_busiest_days busiest_period_of_day  max_order  
0   Regularly busy   Regularly busy        Average orders         10  
1   Regularly busy     Slowest days        Average orders         10  
2   Regularly busy     Slowest days           Most orders         10  
3       Least busy     Slowest days        Average orders         10  
4       Least busy     Slowest days           Most orders         10  
5   Regularly busy   Regularly busy   

### Each value in the "max_order" colum corresponds to the maximum number of orders made by each customer_id. 

In [20]:
ords_prods_merge.head(100)

order_id  customer_id  order_number order_dow  order_hour_of_day  \
0   2539329            1             1         2                  8   
1   2398795            1             2         3                  7   
2    473747            1             3         3                 12   
3   2254736            1             4         4                  7   
4    431534            1             5         4                 15   
..      ...          ...           ...       ...                ...   
95  3226575          360             1         5                 12   
96  1469869          377             3         5                 17   
97  1927023          387             2         4                 10   
98   858092          420             4         1                 19   
99   119314          420             5         3                 19   

    days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                      0.0         196                  1          0   both   
1                     15.0         196                  1          1   both   
2                     21.0         196                  1          1   both   
3                     29.0         196                  1          1   both   
4                     28.0         196                  1          1   both   
..                     ...         ...                ...        ...    ...   
95                     0.0         196                  1          0   both   
96                     3.0         196                  9          0   both   
97                    22.0         196                  3          0   both   
98                    30.0         196                  2          0   both   
99                    16.0         196                  2          1   both   

   product_name  aisle_id  department_id  prices    price_range_loc  \
0          Soda        77              7     9.0  Mid-range product   
1          Soda        77              7     9.0  Mid-range product   
2          Soda        77              7     9.0  Mid-range product   
3          Soda        77              7     9.0  Mid-range product   
4          Soda        77              7     9.0  Mid-range product   
..          ...       ...            ...     ...                ...   
95         Soda        77              7     9.0  Mid-range product   
96         Soda        77              7     9.0  Mid-range product   
97         Soda        77              7     9.0  Mid-range product   
98         Soda        77              7     9.0  Mid-range product   
99         Soda        77              7     9.0  Mid-range product   

       busiest_day new_busiest_days busiest_period_of_day  max_order  
0   Regularly busy   Regularly busy        Average orders         10  
1   Regularly busy     Slowest days        Average orders         10  
2   Regularly busy     Slowest days           Most orders         10  
3       Least busy     Slowest days        Average orders         10  
4       Least busy     Slowest days           Most orders         10  
..             ...              ...                   ...        ...  
95  Regularly busy   Regularly busy           Most orders          3  
96  Regularly busy   Regularly busy           Most orders          3  
97      Least busy     Slowest days           Most orders          8  
98  Regularly busy     Busiest days        Average orders         22  
99  Regularly busy     Slowest days        Average orders         22  

[100 rows x 19 columns]

In [21]:
#Change the number of visible rows
#pd.options.display.max_rows = None

# Deriving columns with loc() function

In [23]:
#Create a flag that assigns a loyalty label based on max order value 

In [24]:
#Loyal customer: max orders is greater than 40
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [25]:
#Regular customers: max orders is greater than 10 but less than 40
ords_prods_merge.loc[(ords_prods_merge['max_order'] >10) & (ords_prods_merge['max_order'] <=40), 'loyalty_flag'] = 'Regular customer'

In [26]:
#New customers: max orders is less than 10
ords_prods_merge.loc[ords_prods_merge['max_order'] <=10, 'loyalty_flag'] = 'New customer'

In [27]:
#Show frequency of each loyalty flag
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [28]:
#Check the dataframe by limiting the output of the head() function to just htose columns you want to check 
ords_prods_merge[['customer_id', 'loyalty_flag', 'order_number']].head(60)
#syntax for selecting specific columns to display

customer_id      loyalty_flag  order_number
0             1      New customer             1
1             1      New customer             2
2             1      New customer             3
3             1      New customer             4
4             1      New customer             5
5             1      New customer             6
6             1      New customer             7
7             1      New customer             8
8             1      New customer             9
9             1      New customer            10
10           15  Regular customer            15
11           15  Regular customer            17
12           15  Regular customer            18
13           15  Regular customer            21
14           15  Regular customer            22
15           19      New customer             2
16           19      New customer             5
17           19      New customer             7
18           21  Regular customer            10
19           31  Regular customer            10
20           31  Regular customer            17
21           43  Regular customer             6
22           43  Regular customer             9
23           52  Regular customer             1
24           52  Regular customer             2
25           52  Regular customer             3
26           52  Regular customer             5
27           52  Regular customer             7
28           52  Regular customer             9
29           52  Regular customer            10
30           52  Regular customer            11
31           52  Regular customer            12
32           52  Regular customer            15
33           52  Regular customer            17
34           52  Regular customer            18
35           52  Regular customer            19
36           52  Regular customer            21
37           67  Regular customer             1
38           67  Regular customer             2
39           67  Regular customer             3
40           67  Regular customer             4
41           67  Regular customer             5
42           67  Regular customer             6
43           67  Regular customer             8
44           67  Regular customer             9
45           67  Regular customer            10
46           67  Regular customer            11
47           67  Regular customer            12
48           67  Regular customer            13
49           67  Regular customer            14
50           67  Regular customer            16
51           67  Regular customer            18
52           67  Regular customer            19
53           67  Regular customer            21
54           67  Regular customer            22
55           67  Regular customer            23
56           81      New customer             3
57           81      New customer             5
58           82  Regular customer             1
59           82  Regular customer             3